HW2/P2/parallel_vector.pyx

```diff
@@ -81,10 +81,14 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0],num_threads=4):
                if counts[src[idx]] > 0:
```

Comment:
How are you ensuring that counts[src[idx]] > 0 continues to be met? With your locking strategy, it is possible for this to eventually be violated.
e.g. say counts[2] = 1 and 2 threads are operating on that same entry. If they both reach (and pass the condition), then they will end up setting counts[2] = -1 once they are both done.
The same comment also applies to your medium-grained locking below.